design/service-binding.md

```diff
@@ -0,0 +1,356 @@
+
+The Service Binding Operator is responsible for binding services such as databases and message brokers to runtime applications in Kubernetes. It's still under development, but the intention is that it becomes part of Operator Lifecycle Management.
+
+## Service Binding Operator today
+
+Today, Strimzi does not fit very nicely with service binding. With the current Service Binding Opeerator implementation, you need to include a template in the `ServiceBindingRequest` to extract the address information from the listener status.
+
+``` yaml
+apiVersion: apps.openshift.io/v1alpha1
+kind: ServiceBindingRequest
+metadata:
+  name: barista-kafka
+spec:
+  applicationSelector:
+    group: apps
+    resource: deployments
+    resourceRef: barista-kafka
+    version: v1
+  backingServiceSelector:
+    group: kafka.strimzi.io
+    kind: Kafka
+    resourceRef: my-cluster
+    version: v1beta1
+  customEnvVar:
+    - name: KAFKA_BOOTSTRAP_SERVERS
+      value: |-
+        {{- range .status.listeners -}}
+          {{- if and (eq .type "plain") (gt (len .addresses) 0) -}}
+            {{- with (index .addresses 0) -}}
+              {{ .host }}:{{ .port }}
+            {{- end -}}
+          {{- end -}}
+        {{- end -}}
+```
+
+While this is quite a clever use of a Go template, it has several problems.
+
+1) The code is included in an annotation for the user's custom resource. This is fragile and ugly.
+1) The code includes the name of a listener. If a different listener name is being used, the code is incorrect.
+1) The code picks just the first of the array of addresses. It's common Kafka practice to have a list of bootstrap servers.
+
+What would be better is to annotate the Strimzi objects in a way that enabled the Service Binding Operator to populate the environment variables itself.
+
+## Service Binding Specification
+
+As an enhancement to the Service Binding Operator, the [Service Binding Specification](https://github.com/application-stacks/service-binding-specification) is being developed by the community to describe how services can be made bindable, which essentially means providing binding data and a description of where to find it.
+
+Using the service binding specification annotations with Strimzi is still a bit fiddly. Here's an example of a `Kafka` CR annotated to enable service binding.
+
+``` yaml
+apiVersion: kafka.strimzi.io/v1beta1
+kind: Kafka
+metadata:
+  name: my-cluster
+  annotations:
+    servicebinding/secret/host: |-
+      {{- range .status.listeners -}}
+        {{- if and (eq .type "plain") (gt (len .addresses) 0) -}}
+          {{- with (index .addresses 0) -}}
+            {{ .host }}
+          {{- end -}}
+        {{- end -}}
+    {{- end -}}
+    servicebinding/secret/port: |-
+      {{- range .status.listeners -}}
+        {{- if and (eq .type "plain") (gt (len .addresses) 0) -}}
+          {{- with (index .addresses 0) -}}
+            {{ .port }}
+          {{- end -}}
+        {{- end -}}
+    {{- end -}}
```

Comment:
I would be for not filling this information here but using a ConfigMap or Secret (as an alternative way provided by the Service Binding specification); I think this way makes more confusion in the Kafka resource which is not related to the binding itself. Anyway, the ConfigMap or Secret creation could be switched on/off based on a env var on the CO.

Reply:
My understanding of this part of the document is that this is what would be required if the Strimzi CRDs were not updated. It basically requires the user to do the hard work of telling the service binding operator where to get credentials. If Strimzi provided a ConfigMap or Secret then this section could be simpler but it would still require the user to add it. Ideally we would update the Strimzi CRDs to include this information (in the simpler form of a ConfigMap or Secret pointer) and then the user does not need to add annotations here.

Reply:
@ppatierno I've heavily revised the proposal now following comments. The Service Binding Operator will actually be able to create its own secret for the binding based on annotations that it reads on the CSV.